File: packages/sdkwork-utils-php/src/StringUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class StringUtils
{
    public static function isBlank(?string $value): bool
    {
        return $value === null || trim($value) === '';
    }

    public static function trim(?string $value): string
    {
        return trim($value ?? '');
    }

    public static function truncate(string $value, int $maxLen, string $suffix = '...'): string
    {
        if ($maxLen <= 0) {
            return '';
        }
        if (mb_strlen($value) <= $maxLen) {
            return $value;
        }
        if (mb_strlen($suffix) >= $maxLen) {
            return mb_substr($suffix, 0, $maxLen);
        }
        return mb_substr($value, 0, $maxLen - mb_strlen($suffix)) . $suffix;
    }

    public static function capitalize(string $value): string
    {
        if ($value === '') {
            return '';
        }
        return strtoupper($value[0]) . strtolower(substr($value, 1));
    }

    public static function camelCase(string $value): string
    {
        $parts = self::camelParts($value);
        if ($parts === []) {
            return '';
        }
        return $parts[0] . implode('', array_map([self::class, 'capitalize'], array_slice($parts, 1)));
    }

    public static function snakeCase(string $value): string
    {
        return implode('_', self::camelParts($value));
    }

    public static function kebabCase(string $value): string
    {
        return implode('-', self::camelParts($value));
    }

    public static function slugify(string $value): string
    {
        return trim(preg_replace('/[^a-z0-9-]/', '', self::kebabCase($value)) ?? '', '-');
    }

    public static function mask(string $value, int $visibleStart, int $visibleEnd, string $maskChar = '*'): string
    {
        if ($visibleStart + $visibleEnd >= strlen($value)) {
            return $value;
        }
        $hidden = strlen($value) - $visibleStart - $visibleEnd;
        return substr($value, 0, $visibleStart) . str_repeat($maskChar, $hidden) . substr($value, -$visibleEnd);
    }

    public static function padStart(string $value, int $targetLen, string $padChar = ' '): string
    {
        if (mb_strlen($value) >= $targetLen) {
            return $value;
        }
        return str_repeat($padChar, $targetLen - mb_strlen($value)) . $value;
    }

    public static function padEnd(string $value, int $targetLen, string $padChar = ' '): string
    {
        if (mb_strlen($value) >= $targetLen) {
            return $value;
        }
        return $value . str_repeat($padChar, $targetLen - mb_strlen($value));
    }

    public static function startsWith(string $value, string $prefix): bool
    {
        return str_starts_with($value, $prefix);
    }

    public static function endsWith(string $value, string $suffix): bool
    {
        return str_ends_with($value, $suffix);
    }

    public static function contains(string $value, string $substring): bool
    {
        return str_contains($value, $substring);
    }

    public static function replaceAll(string $value, string $search, string $replacement): string
    {
        return str_replace($search, $replacement, $value);
    }

    public static function split(string $value, string $delimiter, bool $trimParts = true): array
    {
        $parts = explode($delimiter, $value);
        if (!$trimParts) {
            return $parts;
        }

        $result = [];
        foreach ($parts as $part) {
            $text = trim($part);
            if ($text !== '') {
                $result[] = $text;
            }
        }
        return $result;
    }

    public static function join(array $parts, string $separator): string
    {
        return implode($separator, $parts);
    }

    public static function repeat(string $value, int $count): string
    {
        if ($count < 0) {
            throw new \InvalidArgumentException('repeat count must be >= 0');
        }
        return str_repeat($value, $count);
    }

    public static function normalizeWhitespace(string $value): string
    {
        return trim(preg_replace('/\s+/u', ' ', trim($value)) ?? '');
    }

    public static function template(string $pattern, array $values): string
    {
        return preg_replace_callback(
            '/\{([a-zA-Z_][a-zA-Z0-9_]*)\}/',
            static function (array $matches) use ($values): string {
                return $values[$matches[1]] ?? $matches[0];
            },
            $pattern,
        ) ?? $pattern;
    }

    private static function camelParts(string $value): array
    {
        $normalized = preg_replace('/([a-z0-9])([A-Z])|([A-Z]+)([A-Z][a-z])/', '$1$3 $2$4', trim($value)) ?? '';
        $parts = preg_split('/[^a-zA-Z0-9]+/', $normalized) ?: [];
        return array_values(array_filter(array_map('strtolower', $parts)));
    }
}

File: packages/sdkwork-utils-csharp/StringUtils.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Sdkwork.Utils;

public static partial class StringUtils
{
    public static bool IsBlank(string? value) => string.IsNullOrWhiteSpace(value);

    public static string Trim(string? value) => value?.Trim() ?? string.Empty;

    public static string Truncate(string value, int maxLen, string suffix = "...")
    {
        if (maxLen <= 0) return string.Empty;
        var runes = value.EnumerateRunes().ToArray();
        if (runes.Length <= maxLen) return value;
        var suffixRunes = suffix.EnumerateRunes().ToArray();
        if (suffixRunes.Length >= maxLen)
            return RunesToString(suffixRunes.AsSpan()[..maxLen]);
        var keep = maxLen - suffixRunes.Length;
        return RunesToString(runes.AsSpan()[..keep]) + suffix;
    }

    public static string Capitalize(string value)
    {
        if (string.IsNullOrEmpty(value)) return string.Empty;
        return char.ToUpperInvariant(value[0]) + value[1..].ToLowerInvariant();
    }

    public static string CamelCase(string value)
    {
        var parts = CamelParts(value).ToList();
        if (parts.Count == 0) return string.Empty;
        return parts[0] + string.Concat(parts.Skip(1).Select(Capitalize));
    }

    public static string SnakeCase(string value) => string.Join('_', CamelParts(value));
    public static string KebabCase(string value) => string.Join('-', CamelParts(value));

    public static string Slugify(string value) =>
        Regex.Replace(KebabCase(value), "[^a-z0-9-]", string.Empty).Trim('-');

    public static string Mask(string value, int visibleStart, int visibleEnd, char maskChar = '*')
    {
        var runes = value.EnumerateRunes().ToArray();
        if (visibleStart + visibleEnd >= runes.Length) return value;
        return RunesToString(runes.AsSpan()[..visibleStart])
            + new string(maskChar, runes.Length - visibleStart - visibleEnd)
            + RunesToString(runes.AsSpan()[^visibleEnd..]);
    }

    public static string PadStart(string value, int targetLen, char padChar = ' ')
    {
        var current = value.EnumerateRunes().Count();
        if (current >= targetLen) return value;
        return new string(padChar, targetLen - current) + value;
    }

    public static string PadEnd(string value, int targetLen, char padChar = ' ')
    {
        var current = value.EnumerateRunes().Count();
        if (current >= targetLen) return value;
        return value + new string(padChar, targetLen - current);
    }

    public static bool StartsWith(string value, string prefix) => value.StartsWith(prefix);

    public static bool EndsWith(string value, string suffix) => value.EndsWith(suffix);

    public static bool Contains(string value, string substring) => value.Contains(substring);

    public static string ReplaceAll(string value, string search, string replacement) =>
        value.Replace(search, replacement);

    public static List<string> Split(string value, string delimiter, bool trimParts = true)
    {
        var parts = value.Split(delimiter);
        var result = new List<string>(parts.Length);
        foreach (var part in parts)
        {
            var text = trimParts ? part.Trim() : part;
            if (!trimParts || text.Length > 0) result.Add(text);
        }
        return result;
    }

    public static string Join(IEnumerable<string> parts, string separator) => string.Join(separator, parts);

    public static string Repeat(string value, int count)
    {
        if (count < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(count), "repeat count must be >= 0");
        }

        return string.Concat(Enumerable.Repeat(value, count));
    }

    public static string NormalizeWhitespace(string value) =>
        string.Join(' ', value.Trim().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));

    private static readonly System.Text.RegularExpressions.Regex TemplateKey =
        new(@"\{([a-zA-Z_][a-zA-Z0-9_]*)\}", System.Text.RegularExpressions.RegexOptions.Compiled);

    public static string Template(string pattern, IReadOnlyDictionary<string, string> values) =>
        TemplateKey.Replace(pattern, match =>
        {
            var key = match.Groups[1].Value;
            return values.TryGetValue(key, out var replacement) ? replacement : match.Value;
        });

    private static IEnumerable<string> CamelParts(string value)
    {
        var normalized = CamelBoundary().Replace(value.Trim(), "$1$3 $2$4");
        return WordSplit().Split(normalized).Where(part => part.Length > 0).Select(part => part.ToLowerInvariant());
    }

    [GeneratedRegex("[^a-zA-Z0-9]+")]
    private static partial Regex WordSplit();

    [GeneratedRegex("([a-z0-9])([A-Z])|([A-Z]+)([A-Z][a-z])")]
    private static partial Regex CamelBoundary();

    private static string RunesToString(ReadOnlySpan<Rune> runes)
    {
        var builder = new StringBuilder();
        Span<char> buffer = stackalloc char[2];
        foreach (var rune in runes)
        {
            var length = rune.EncodeToUtf16(buffer);
            builder.Append(buffer[..length]);
        }
        return builder.ToString();
    }
}
